src/TLE.ts

```diff
@@ -874,65 +874,95 @@ export class Parser {
         assert(1 <= stringValue.length, "TLE strings must be at least 1 character.");
         assert(Utilities.isQuoteCharacter(stringValue[0]), "The first character in a TLE string must be a quote character.");
 
-        let leftSquareBracketToken: Token = null;
+        const leftSquareBracketToken: Token = null;
         let expression: Value = null;
-        let rightSquareBracketToken: Token = null;
-        let errors: language.Issue[] = [];
+        const rightSquareBracketToken: Token = null;
+        const errors: language.Issue[] = [];
+        let result: { stringValue: string; leftSquareBracketToken: Token; expression: Value; rightSquareBracketToken: Token; errors: language.Issue[] } = { stringValue, leftSquareBracketToken, expression, rightSquareBracketToken, errors };
 
         if (3 <= stringValue.length && stringValue.substr(1, 2) === "[[") {
+            result.expression = new StringValue(Token.createQuotedString(0, stringValue));
+        } else if (stringValue[1] === '[' && stringValue[stringValue.length - 2] === ']') {
+            if (3 <= stringValue.length && stringValue.substr(1, 2) === "[[") {
+                expression = new StringValue(Token.createQuotedString(0, stringValue));
+            } else {
+                result = Parser.parseStringAsExpression(stringValue, leftSquareBracketToken, expression, rightSquareBracketToken, errors);
+            }
+        } else {
+            let stringCase = false;
+            let index = 0;
+            if (stringValue[1] === '[') {
+                for (let rightBracket of stringValue) {
+                    ++index;
+                    if (rightBracket === ']') {
+                        for (index; index < stringValue.length; ++index) {
+                            if (stringValue[index] === '"') {
+                                result.expression = new StringValue(Token.createQuotedString(0, stringValue));
+                                stringCase = true;
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+            if (stringCase === false) {
+                result = Parser.parseStringAsExpression(stringValue, leftSquareBracketToken, expression, rightSquareBracketToken, errors);
+            }
+        }
+        return new ParseResult(result.leftSquareBracketToken, result.expression, result.rightSquareBracketToken, result.errors);
+    }
+
+    private static parseStringAsExpression(stringValue: string, leftSquareBracketToken: Token, expression: Value, rightSquareBracketToken: Token, errors: language.Issue[]): { stringValue: string; leftSquareBracketToken: Token; expression: Value; rightSquareBracketToken: Token; errors: language.Issue[] } {
```

Comment:
I'm not convinced you've gained anything by refactoring out praseStringAsExpression.  It seems like it's only made it more complex and harder to read, especially with the way you're passing parameters in and out.

I really don't think all the changes you have are necessary.  I would start again with the original function for this file and make just the changes necessary.

Reply:
Again look at my suggested if/then.  I think it really should be as easy as getting rid of the current "if (3 <= strintgValue.length…" statement and replacing it with my suggestion:

   if (
        stringValue[1] !== "[" // Does not start with '['
        || stringValue.substr(1, 2) === "[[" // Starts with '[['
        || (stringValue[1] === '[' && stringValue[stringValue.length - 1] === ']') // Starts with '[' but doesn't end with ']'
    ) {
        // This is considered a plain string
        expression = new StringValue(Token.createQuotedString(0, stringValue));
    } else {
      // This is considered an expression
      let tokenizer = Tokenizer.fromString(stringValue);
      tokenizer.next();
      ...